test/e2e/runner/evidence.go

```diff
@@ -0,0 +1,265 @@
+package main
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io/ioutil"
+	"math/rand"
+	"path/filepath"
+	"time"
+
+	"github.com/tendermint/tendermint/crypto"
+	"github.com/tendermint/tendermint/crypto/tmhash"
+	tmjson "github.com/tendermint/tendermint/libs/json"
+	"github.com/tendermint/tendermint/privval"
+	tmproto "github.com/tendermint/tendermint/proto/tendermint/types"
+	e2e "github.com/tendermint/tendermint/test/e2e/pkg"
+	"github.com/tendermint/tendermint/types"
+	"github.com/tendermint/tendermint/version"
+)
+
+// 1 in 4 evidence is light client evidence, the rest is duplicate vote
+const lightClientEvidenceRatio = 4
+
+// InjectEvidence takes a running testnet and generates an amount of valid
+// evidence to a random node through the rpc endpoint `/broadcast_evidence`
+// Evidence is random and can be a mixture of LightClientAttackEvidence and
+// DuplicateVoteEvidence
+func InjectEvidence(testnet *e2e.Testnet, amount int) error {
+	// select a random node
+	targetNode := testnet.RandomNode()
+
+	logger.Info(fmt.Sprintf("Injecting evidence through %v (amount: %d)...", targetNode.Name, amount))
+
+	client, err := targetNode.Client()
+	if err != nil {
+		return err
+	}
+
+	// request the latest block and validator set from the node
+	blockRes, err := client.Block(context.Background(), nil)
+	if err != nil {
+		return err
+	}
+	nValidators := 100
+	valRes, err := client.Validators(context.Background(), &blockRes.Block.Height, nil, &nValidators)
+	if err != nil {
+		return err
+	}
+	valSet, err := types.ValidatorSetFromExistingValidators(valRes.Validators)
+	if err != nil {
+		return err
+	}
+
+	// get the private keys of all the validators in the network
+	privVals, err := getPrivateValidatorKeys(testnet)
+	if err != nil {
+		return err
+	}
+
+	// wait for the node to reach the height above the forged height so that
+	// it is able to validate the evidence
+	_, err = waitForNode(targetNode, blockRes.Block.Height+3, 10*time.Second)
```

Comment:
Why is height+3 used specifically?

Reply:
The light client attack is a lunatic attack that is 2 blocks ahead of the current blockchain (I didn't use 1 because we already know the next validator set). Then I wait an extra block for safety so three in total